distributed/tests/test_client.py

```diff
@@ -7345,6 +7336,109 @@ async def test_dump_cluster_state_json(c, s, a, b, tmp_path, local):
         await c.dump_cluster_state(filename, format="json")
 
 
+def blocked_inc(x, event):
+    event.wait()
+    return x + 1
+
+
+@pytest.mark.parametrize("local", [True, False])
+@pytest.mark.parametrize("_format", ["msgpack", "yaml"])
+@gen_cluster(client=True)
+async def test_inspect_cluster_dump(c, s, a, b, tmp_path, _format, local):
+    filename = tmp_path / "foo"
+    if not local:
+        pytest.importorskip("fsspec")
+        # Make it look like an fsspec path
+        filename = f"file://{filename}"
+
+    futs = c.map(inc, range(2))
+    fut_keys = {f.key for f in futs}
+    await c.gather(futs)
+
+    event = distributed.Event()
+    blocked_fut = c.submit(blocked_inc, 1, event)
+
+    await c.dump_cluster_state(filename, format=_format)
+
+    scheduler_tasks = list(s.tasks.values())
+    worker_tasks = [t for w in (a, b) for t in w.tasks.values()]
+
+    event.set()
+
+    suffix = ".gz" if _format == "msgpack" else ""
+    dump = DumpArtefact(f"{filename}.{_format}{suffix}")
+
+    smem_keys = {t["key"] for t in dump.scheduler_tasks("memory")}
+    wmem_keys = {t["key"] for t in dump.worker_tasks("memory")}
+
+    assert smem_keys == fut_keys
+    assert smem_keys == {t.key for t in scheduler_tasks if t.state == "memory"}
```

Comment:
I'm kinda surprised this works. Because `scheduler_tasks` is just a list of the actual TaskState objects on the scheduler (shallow copy, not deep copy), when `blocked_inc` completes, the TaskState object for `blocked_inc` will mutate and switch from state `processing` to `memory`. But the dump was captured while `blocked_inc` was still `processing`, so then this won't match up.

I bet if you added an `await asyncio.sleep(0.1)` right after `event.set`, this would fail.

You could just move the `event.set()` to the very end of the test. But with such a small example like you have here, I think it would be perfectly reasonable to just hardcode the expected results in this test (pass `key=` to `c.submit` so you can control the task names). Comparing against the actual state of the scheduler is kinda nice, but might be more complex than you need.

Reply:
This is also a good catch. I discovered that the problem was that I wasn't awaiting the `event.set()` which meant that the wait in the `blocked_inc` never woke up.

I've reworked the test to

```python
await asyncio.sleep(0.05)  # Give blocked_inc time to start
await  c.dump_cluster_state(filename, format="msgpack")
...
await c.gather(event.set(), blocked_fut)
```